python3Full: obsolete and remove

mweinelt opened the pull request:
Why? python3Full must die.

Wondering if there is more beautiful way to cobble the source together.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mweinelt committed:
python3Packages.tkinter: build standalone from python source tree

Why? python3Full must die.

emilazy commented on pkgs/development/python-modules/tkinter/default.nix:
```diff
@@ -1,34 +1,62 @@
+    cp -rv Lib/tkinter ../tkinter/
+
+    pushd $NIX_BUILD_TOP/tkinter
+    cp ${./pyproject.toml} ./pyproject.toml
```
You didn’t commit this. But maybe `replaceVars` would be nice here, or even just using `pkgs.formats.toml`.

Maybe we can avoid copying around the source by just using the full paths in `pyproject.toml`? But I guess `setuptools` might expect a different layout.

mweinelt replied:
I don't really want to get too fancy with the pyproject.toml and setuptools, this needs to stay maintainable after all.

LordGrimmauld replied:
You can do the following change:
```diff
diff --git a/pkgs/development/python-modules/tkinter/default.nix b/pkgs/development/python-modules/tkinter/default.nix
index ba4285ca329c..4fdfac369310 100644
--- a/pkgs/development/python-modules/tkinter/default.nix
+++ b/pkgs/development/python-modules/tkinter/default.nix
@@ -9,6 +9,7 @@
   tk,
   tkinter,
   xvfb-run,
+  replaceVars,
 }:
 
 buildPythonPackage {
@@ -18,7 +19,14 @@ buildPythonPackage {
 
   src = python.src;
 
-  prePatch = ''
+  prePatch =
+    let
+      pyproj = replaceVars ./pyproject.toml {
+        python_version = python.version;
+        python_internal_dir = "${python}/include/${python.libPrefix}/internal";
+      };
+    in
+    ''
       mkdir $NIX_BUILD_TOP/tkinter
 
       # copy the module bits and pieces from the python source
@@ -29,11 +37,7 @@ buildPythonPackage {
       pushd $NIX_BUILD_TOP/tkinter
 
       # install our custom pyproject.toml
-    cp ${./pyproject.toml} ./pyproject.toml
-    substituteInPlace pyproject.toml \
-      --replace-fail "@python_version@" "${python.version}" \
-      --replace-fail "@python_internal_dir@" "${python}/include/${python.libPrefix}/internal"
-
+      cp ${pyproj} ./pyproject.toml
     ''
     + lib.optionalString (pythonOlder "3.13") ''
       substituteInPlace "tkinter/tix.py" --replace-fail \
```

(careful: the formatter will have to rerun, this patch is `--ignore-all-space` for better readability)

emilazy reviewed (changes requested):
There are tests in `Lib/test/test_tkinter` and `Lib/test/tkinterdata` that we should probably try to preserve and run.

There are a couple downstream users in CPython: `turtle` (and `turtledemo`) and `idlelib` (including the IDLE application). I guess we could bundle them into this package, but it probably makes more sense to apply the same trick to those as separate module packages. They’re both used a fair amount in education, I think, so it would be best to keep them available unless they’re onerous to package.

emilazy commented on pkgs/development/python-modules/tkinter/default.nix:
```diff
@@ -1,34 +1,62 @@
+    mkdir $NIX_BUILD_TOP/tkinter
+
+    cp -v  Modules/{_tkinter.c,tkinter.h} ../tkinter/
+    cp -rv Modules/clinic ../tkinter/
```
Maybe we should expose the Argument Clinic as an output of the Python derivation or something? They do say it’s internal‐only, but also that “if you maintain an external C extension for CPython, you’re welcome to experiment with Argument Clinic in your own code”.

mweinelt replied:
I don't fancy exposing things for the thrill of it. It wasn't needed until now, so I won't  bother.

emilazy commented on pkgs/development/python-modules/tkinter/default.nix:
```diff
@@ -1,34 +1,66 @@
+      "-I${lib.getDev tcl}/include"
+      "-I${lib.getDev tk}/include"
+    ];
+  };
```
This behaves weirdly with various position attributes. Better to `inherit (python.meta) …;` (but as it’s pre-existing, not a blocker of course).

mweinelt commented:
> There are tests in Lib/test/test_tkinter and Lib/test/tkinterdata that we should probably try to preserve and run.

These are really annoying to get to run, so I'd rather not.

> There are a couple downstream users in CPython: turtle (and turtledemo) and idlelib (including the IDLE application). I guess we could bundle them into this package, but it probably makes more sense to apply the same trick to those as separate module packages. They’re both used a fair amount in education, I think, so it would be best to keep them available unless they’re onerous to package.

They are importable with standard python that has tkinter in its python path.

LordGrimmauld reviewed (commented):
Implementation looks sane.
`tkinter` does seem to work, at least i can import it and launch a window. Test fails don't look nice, but this is probably better than before, where we had no tests at all.

mweinelt commented on pkgs/development/python-modules/tkinter/default.nix:
```diff
@@ -1,34 +1,86 @@
 {
```
I think we duplicate the library code between python stdlib and the site-package now. I don't care enough.

mweinelt commented on pkgs/development/python-modules/tkinter/default.nix:
```diff
@@ -1,34 +1,86 @@
+
+    runHook postCheck
+  '';
+
```
Added.

mweinelt committed:
cpython: remove x11support

This is required for tkinter, which we now build standalone as tkinter
in the python package set.

LordGrimmauld commented:
Needs to target staging, other than that looks reasonable enough

mweinelt commented:
Last call.

LordGrimmauld reviewed (dismissed):
tkinter changes work locally, diff on py full looks reasonable but i didn't get to build it yet

LordGrimmauld reviewed (commented):
Uh - `idle` from `python3Full` is now broken:
```
** IDLE can't import Tkinter.
Your Python may not be configured for Tk. **
```
We might have to add tkinter back to it for idle to work?

mweinelt commented:
Yeah, you now need to include tkinter in the python environment.

mweinelt committed:
bluez-headers: expose bluez header files

This escapes infinite recursion, because Python cannot depend on bluez
due to infinite recursion on multiple paths. But it only wants
bluetooth.h at build time, so this is sufficient.

In 5.84 the source path changes from lib/ to lib/bluetooth, which will
require an update to the install phase.

mweinelt committed:
cpython: build with bluetooth support in default build

Closes: #341001

mweinelt committed:
openfreebuds: move to default python

The default python now provides bluetooth support.

mweinelt committed:
python3Full: drop

Bluetooth support is now built into the default build. The X11 support is
realized outside the Python interpreter by including the tkinter module
from the package set.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/94

mweinelt commented:
Tested on aarch64-linux and x86_64-linux.

Still has an infinite recursion on cross.

SuperSandro2000 reviewed (approved):
big fan of this

mweinelt commented:
I think I found the way that sucks least.

mweinelt commented on pkgs/by-name/bl/bluez-headers/package.nix:
```diff
@@ -0,0 +1,24 @@
+{
+  stdenv,
+  bluez,
+  ...
```
Dropped.

LordGrimmauld commented on pkgs/by-name/bl/bluez-headers/package.nix:
```diff
@@ -0,0 +1,42 @@
+      lgpl21Plus
+      mit
+    ];
+    mainProgram = "btinfo";
```
```suggestion
```
`mainProgram` on a headers-only package is silly

LordGrimmauld commented on pkgs/by-name/bl/bluez/package.nix:
```diff
@@ -186,18 +182,5 @@ stdenv.mkDerivation (finalAttrs: {
-    maintainers = with lib.maintainers; [ ];
-    platforms = lib.platforms.linux;
-  };
+  inherit (bluez-headers) meta;
```
```suggestion
  meta = bluez-headers.meta // { mainProgram = "btinfo"; };
```

mweinelt commented:
Good catch, applied.

mweinelt commented:
Added another comment why the bluez-headers package has the src fod and not bluez.

dotlambda commented on pkgs/by-name/bl/bluez/package.nix:
```diff
@@ -186,18 +182,7 @@ stdenv.mkDerivation (finalAttrs: {
-      lgpl21Plus
-      mit
-    ];
+  meta = bluez-headers.meta // {
```
This breaks `meta.position`.
